ruby_on_rails/testing/model_testing.md

```diff
@@ -0,0 +1,125 @@
+end
+
+~~~
+
+Make sure to require this file in `rails_helper.rb`. We will briefly cover what a factory will look like and some things to keep in mind when creating your data. While you can require the file by adding `require support/factory_bot` to the top of the file. Rails also provides this in the commented out line `Dir[Rails.root.join("spec/support/**/*.rb")].each { |f| require f }` which enables you to load all of the support files when booting up. This is useful for not having to manually add all of the support files and avoids issues if you forget to add a file for example. 
+
+The last thing you might be wondering is how the test database clears itself in between it's tests. Rail's does this automatically for you, however if Rail's is not doing it for you make sure your following the instructions in the rails_helper.rb file and make sure to ask in the [Discord](https://discord.com/channels/505093832157691914/690591236922409012) first! There is also a alternative option in the additional resources section if need be for database cleaning.
+
+## Creating a factory
+
+Below is an example of a user factory. Notice that the factory will be a User class based off of the name of your factory. You can also define as many attributes as you want and it can be useful to focus on adding only the attributes your testing and adding them as need be.
+
+~~~
+
+# This will guess the User class
+FactoryBot.define do
+  factory :user do
+    first_name { "John" }
+    last_name  { "Doe" }
+    admin { false }
+  end
+end
+
+~~~
+
+When creating a factory bot record you have the choice of using the `create(:user)` which returns a saved instance of the object from your database or `build(:user)` which returns an unsaved instance of an object. 
+
+## Example model test
+
+In the following example one of the things we often see when users sign up is that sites will ask for an email to send a user notifications for example. RSpec can help you test this by utilizing model tests. Let's take a look at what a requiring a user email would look like in a model and the corresponding tests. 
+
+~~~
+
+class EmailValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    unless value =~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
+      record.errors.add attribute, (options[:message] || "is not an email")
+    end
+  end
+end
+
+class User < ApplicationRecord
+  validates :email, presence: true, email: true
+end
+
+~~~
+ 
+In the above code we add our own validator to the email validation. [Validate each](https://guides.rubyonrails.org/active_record_validations.html#custom-validators). By using validate the model will call on the validator to ensure the presence of valid characters. That then brings us to the following RSpec test where we will write the context and what we expect the outcome to be. The validate_each method is also where you would make changes as well if the test failed to pass.
```

Comment:
Really love this idea! However, the test is truly just testing to see if the email address was provided, so I do not think we need to add the `EmailValidator` logic to this testing lesson. Though this would be great information in a lesson!  I feel that it is distracting from the main purpose of this lesson. I was imagining just a test for the `validates :email, presence: true` portion of the `User` model.

If you feel strongly about the need for testing an email also means that we need to test that it is a valid email address, then I suggest that we skip this idea and go with an easier validation for `validates : first_name, presence: true`